src/handle.rs

```diff
@@ -432,3 +449,301 @@ impl HandleHost {
     unsafe { self.get_isolate().as_ref() }.thread_safe_handle()
   }
 }
+
+/// An object reference that does not prevent garbage collection for the object,
+/// and which allows installing finalization callbacks which will be called
+/// after the object has been GC'd.
+///
+/// Note that finalization callbacks are tied to the lifetime of a `Weak<T>`,
+/// and will not be called after the `Weak<T>` is dropped.
+///
+/// # `Clone`
+///
+/// Since finalization callbacks are specific to a `Weak<T>` instance, cloning
+/// will create a new object reference without a finalizer, as if created by
+/// [`Self::new`]. You can use [`Self::clone_with_finalizer`] to attach a
+/// finalization callback to the clone.
+#[derive(Debug)]
+pub struct Weak<T> {
+  data: Option<Box<WeakData<T>>>,
+  isolate_handle: IsolateHandle,
+}
+
+impl<T> Weak<T> {
```

Comment:
Talked with @littledivy and it seems we're gonna need similar API to `v8::Global` to be able to turn `Weak` into raw pointer and then back into `v8::Weak` - maybe we should add public APIs `into_raw()` and `from_raw()`?

Reply:
Assuming it doesn't matter for the NAPI that the pointer points to a `WeakData<T>` rather than pointing into V8's heap, I guess that's fine.

Reply:
> Assuming it doesn't matter for the NAPI that the pointer points to a `WeakData<T>` in the Rust heap rather than pointing into V8's heap, I guess that's fine.

Yeah, I think the only limitation is that it's expected that it's a size of a single pointer, so we'd need to manually pass back `Isolate` - but the pattern from `v8::Global` works fine now, so we should be good.